File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.py
import sys
input = sys.stdin.readline
from collections import deque

n,m = map(int,input().split())
board = [list(input().rstrip()) for _ in range(m)]

dx = [-1,1,0,0]
dy = [0,0,-1,1]

def bfs(i,j,color):
    queue = deque([])
    queue.append([i,j])
    cnt = 0
    board[i][j] = 0

    while queue:
        x,y = queue.popleft()
        cnt += 1
        for k in range(4):
            nx = x+dx[k]
            ny = y+dy[k]
            if 0<=nx<m and 0<=ny<n:
                if board[nx][ny] == color:
                    board[nx][ny] = 0
                    queue.append([nx,ny])
                else:
                    continue
    
    return cnt

w,b =0,0
for i in range(m):
    for j in range(n):
        if board[i][j] == 'W':
            w += (bfs(i,j,'W'))**2
        elif board[i][j] == 'B':
            b += (bfs(i,j,'B'))**2

print(w,b)

File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.js
const fs = require('fs');
// let input = fs.readFileSync('./input.txt').toString().trim().split('\n');

/**정답 제출 시에는 이 코드로 제출 */
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

let [n,m] = input.shift().split(' ').map(Number);
let board = input.slice(0).map((v) => v.trim().split(""));
let dx = [-1,1,0,0];
let dy = [0,0,-1,1];

const bfs = (i,j) => {
  let cnt = 0;
  let queue = [[i,j]];
  let color = board[i][j];
  board[i][j] = 0;

  while (queue.length) {
    let [x,y] = queue.shift();
    cnt += 1;

    for (let k=0; k<4; k++) {
      let nx = x+dx[k]
      let ny = y+dy[k]
      if (0<=nx && nx <m && 0<=ny && ny<n && color === board[nx][ny]) {
        board[nx][ny] = 0;
        queue.push([nx,ny]);
      }
    }
  }

  color === 'W' ? (w+=cnt**2) : (b+=cnt**2);
}

let w =0;
let b = 0;

for (let i=0; i<m; i++){
  for (let j=0; j<n; j++){
    if (board[i][j]) bfs(i,j)
  }
}

console.log(w + ' ' + b);